{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e428e461",
   "metadata": {},
   "source": [
    "# Preparing the Data for Training\n",
    "\n",
    "### This Jupyter Notebook replicated the preprocessing steps on the original paper for better understanding\n",
    "\n",
    "The architecture of DeepSignalingFlow forces us to create seperate networks for each available dataset depending on the number of genes and drug combinations. We will thus start by only preprocessing the NCI ALMANAC Dataset.\n",
    "\n",
    "To prepare we need to download the following Datasets:\n",
    "\n",
    "NCI ALMANAC [Download](https://wiki.nci.nih.gov/spaces/NCIDTPdata/pages/338237347/NCI-ALMANAC) <br>\n",
    "KEGG [Download](https://www.genome.jp/kegg/) <br>\n",
    "DrugBank [Download](https://go.drugbank.com/) <br>\n",
    "Cell Model Passports RNA-Seq [Download](https://cog.sanger.ac.uk/cmp/download/rnaseq_20191101.zip) <br>\n",
    "Cell Model Passports CNV [Download](https://cog.sanger.ac.uk/cmp/download/cnv_20191101.zip) <br>\n",
    "\n",
    "NOTES: \n",
    "We do not know which KEGG dataset has been used here.\n",
    "The NCI ALMANAC dataset is called the NCI60 in the code which is not the same.\n",
    "\n",
    "\n",
    "#### Reading our dataset\n",
    "Now we begin by loading our dataset and editing it into a new file.\n",
    "\n",
    "In the original code this can be found at `parse/init_parse.py`\n",
    "\n",
    "Let's do some basic tests and follow their process:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "3b6bbaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "af7b4691",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Michael Migacev\\AppData\\Local\\Temp\\ipykernel_1384740\\1597779311.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  dl_input_df = pd.read_csv('../data-nci/raw_data/nci/nci-raw-data.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COMBODRUGSEQ</th>\n",
       "      <th>SCREENER</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>TESTDATE</th>\n",
       "      <th>PLATE</th>\n",
       "      <th>PANELNBR</th>\n",
       "      <th>CELLNBR</th>\n",
       "      <th>PREFIX1</th>\n",
       "      <th>NSC1</th>\n",
       "      <th>SAMPLE1</th>\n",
       "      <th>...</th>\n",
       "      <th>PERCENTGROWTH</th>\n",
       "      <th>PERCENTGROWTHNOTZ</th>\n",
       "      <th>TESTVALUE</th>\n",
       "      <th>CONTROLVALUE</th>\n",
       "      <th>TZVALUE</th>\n",
       "      <th>EXPECTEDGROWTH</th>\n",
       "      <th>SCORE</th>\n",
       "      <th>VALID</th>\n",
       "      <th>PANEL</th>\n",
       "      <th>CELLNAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>260496</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>85.979</td>\n",
       "      <td>88.159</td>\n",
       "      <td>332168.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>90.342</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>260497</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>100.903</td>\n",
       "      <td>100.763</td>\n",
       "      <td>379656.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>87.130</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>260498</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>14.147</td>\n",
       "      <td>27.498</td>\n",
       "      <td>103608.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>12.739</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>260499</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>71.268</td>\n",
       "      <td>75.736</td>\n",
       "      <td>285360.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>76.397</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>260500</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>89.278</td>\n",
       "      <td>90.945</td>\n",
       "      <td>342664.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>73.681</td>\n",
       "      <td>-16.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   COMBODRUGSEQ SCREENER                STUDY    TESTDATE        PLATE  \\\n",
       "0        260496       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "1        260497       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "2        260498       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "3        260499       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "4        260500       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "\n",
       "   PANELNBR  CELLNBR PREFIX1  NSC1  SAMPLE1  ...  PERCENTGROWTH  \\\n",
       "0         9       18       S   752       37  ...         85.979   \n",
       "1         9       18       S   752       37  ...        100.903   \n",
       "2         9       18       S   752       37  ...         14.147   \n",
       "3         9       18       S   752       37  ...         71.268   \n",
       "4         9       18       S   752       37  ...         89.278   \n",
       "\n",
       "   PERCENTGROWTHNOTZ TESTVALUE CONTROLVALUE    TZVALUE  EXPECTEDGROWTH  SCORE  \\\n",
       "0             88.159  332168.0   376781.548  58592.854          90.342    4.0   \n",
       "1            100.763  379656.0   376781.548  58592.854          87.130  -14.0   \n",
       "2             27.498  103608.0   376781.548  58592.854          12.739   -1.0   \n",
       "3             75.736  285360.0   376781.548  58592.854          76.397    5.0   \n",
       "4             90.945  342664.0   376781.548  58592.854          73.681  -16.0   \n",
       "\n",
       "   VALID         PANEL  CELLNAME  \n",
       "0      Y  Renal Cancer     786-0  \n",
       "1      Y  Renal Cancer     786-0  \n",
       "2      Y  Renal Cancer     786-0  \n",
       "3      Y  Renal Cancer     786-0  \n",
       "4      Y  Renal Cancer     786-0  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dl_input_df = pd.read_csv('../data-nci/raw_data/nci/nci-raw-data.csv')\n",
    "dl_input_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b4c3ab",
   "metadata": {},
   "source": [
    "Now we can copy the first operation from the paper, we can see that given no elements called \"Drug A\" or \"Drug B\" we cannot perform the first operation on the raw data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "220bbfce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dl_input_df = dl_input_df.groupby(['Drug A', 'Drug B', 'Cell Line Name']).agg({'Score':'mean'}).reset_index()\n",
    "#dl_input_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41f19035",
   "metadata": {},
   "source": [
    "To perform the same operations we need to rename the field so they adhere to the raw data. The corresponding fields could be: <br>\n",
    "`NSC1` or <br>\n",
    "`SAMPLE1` <br>\n",
    "This should not matter much as we just want to look at the data for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c2f34567",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(62277, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE1</th>\n",
       "      <th>SAMPLE2</th>\n",
       "      <th>CELLNAME</th>\n",
       "      <th>SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62272</th>\n",
       "      <td>290</td>\n",
       "      <td>31.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62273</th>\n",
       "      <td>290</td>\n",
       "      <td>46.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62274</th>\n",
       "      <td>290</td>\n",
       "      <td>52.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-7.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62275</th>\n",
       "      <td>290</td>\n",
       "      <td>91.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62276</th>\n",
       "      <td>290</td>\n",
       "      <td>203.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>3.222222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE1  SAMPLE2 CELLNAME     SCORE\n",
       "62272      290     31.0   HCT-15 -2.444444\n",
       "62273      290     46.0   HCT-15 -2.000000\n",
       "62274      290     52.0   HCT-15 -7.888889\n",
       "62275      290     91.0   HCT-15 -2.444444\n",
       "62276      290    203.0   HCT-15  3.222222"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dl_input_df = dl_input_df.groupby(['SAMPLE1', 'SAMPLE2', 'CELLNAME']).agg({'SCORE':'mean'}).reset_index()\n",
    "print(dl_input_df.shape)\n",
    "dl_input_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "02cf5b85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(62277, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE1</th>\n",
       "      <th>SAMPLE2</th>\n",
       "      <th>CELLNAME</th>\n",
       "      <th>SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62272</th>\n",
       "      <td>290</td>\n",
       "      <td>31.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62273</th>\n",
       "      <td>290</td>\n",
       "      <td>46.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62274</th>\n",
       "      <td>290</td>\n",
       "      <td>52.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-7.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62275</th>\n",
       "      <td>290</td>\n",
       "      <td>91.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62276</th>\n",
       "      <td>290</td>\n",
       "      <td>203.0</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>3.222222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE1  SAMPLE2 CELLNAME     SCORE\n",
       "62272      290     31.0   HCT-15 -2.444444\n",
       "62273      290     46.0   HCT-15      -2.0\n",
       "62274      290     52.0   HCT-15 -7.888889\n",
       "62275      290     91.0   HCT-15 -2.444444\n",
       "62276      290    203.0   HCT-15  3.222222"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dl_input_deletion_list = []\n",
    "dl_input_df = dl_input_df.fillna('missing')\n",
    "for row in dl_input_df.itertuples():\n",
    "    if row[1] == 'missing' or row[2] == 'missing':\n",
    "        dl_input_deletion_list.append(row[0])\n",
    "dl_input_df = dl_input_df.drop(dl_input_df.index[dl_input_deletion_list]).reset_index(drop=True)\n",
    "print(dl_input_df.shape)\n",
    "dl_input_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22454c80",
   "metadata": {},
   "source": [
    "Something seems wrong here. Not only are there some preprocessing operations not mentioned but no missing values are getting removed, this may result in the 2 different datatypes of `SAMPLE1` and `SAMPLE2` <br>\n",
    "So Let's investigate a bit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "d2510937",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Michael Migacev\\AppData\\Local\\Temp\\ipykernel_1384740\\1597779311.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  dl_input_df = pd.read_csv('../data-nci/raw_data/nci/nci-raw-data.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COMBODRUGSEQ</th>\n",
       "      <th>SCREENER</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>TESTDATE</th>\n",
       "      <th>PLATE</th>\n",
       "      <th>PANELNBR</th>\n",
       "      <th>CELLNBR</th>\n",
       "      <th>PREFIX1</th>\n",
       "      <th>NSC1</th>\n",
       "      <th>SAMPLE1</th>\n",
       "      <th>...</th>\n",
       "      <th>PERCENTGROWTH</th>\n",
       "      <th>PERCENTGROWTHNOTZ</th>\n",
       "      <th>TESTVALUE</th>\n",
       "      <th>CONTROLVALUE</th>\n",
       "      <th>TZVALUE</th>\n",
       "      <th>EXPECTEDGROWTH</th>\n",
       "      <th>SCORE</th>\n",
       "      <th>VALID</th>\n",
       "      <th>PANEL</th>\n",
       "      <th>CELLNAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>260496</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>85.979</td>\n",
       "      <td>88.159</td>\n",
       "      <td>332168.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>90.342</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>260497</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>100.903</td>\n",
       "      <td>100.763</td>\n",
       "      <td>379656.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>87.130</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>260498</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>14.147</td>\n",
       "      <td>27.498</td>\n",
       "      <td>103608.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>12.739</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>260499</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>71.268</td>\n",
       "      <td>75.736</td>\n",
       "      <td>285360.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>76.397</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>260500</td>\n",
       "      <td>FG</td>\n",
       "      <td>PZUT00156_86_01_T72</td>\n",
       "      <td>06/23/2011</td>\n",
       "      <td>786-0_1_T72</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>S</td>\n",
       "      <td>752</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>89.278</td>\n",
       "      <td>90.945</td>\n",
       "      <td>342664.0</td>\n",
       "      <td>376781.548</td>\n",
       "      <td>58592.854</td>\n",
       "      <td>73.681</td>\n",
       "      <td>-16.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Renal Cancer</td>\n",
       "      <td>786-0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   COMBODRUGSEQ SCREENER                STUDY    TESTDATE        PLATE  \\\n",
       "0        260496       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "1        260497       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "2        260498       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "3        260499       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "4        260500       FG  PZUT00156_86_01_T72  06/23/2011  786-0_1_T72   \n",
       "\n",
       "   PANELNBR  CELLNBR PREFIX1  NSC1  SAMPLE1  ...  PERCENTGROWTH  \\\n",
       "0         9       18       S   752       37  ...         85.979   \n",
       "1         9       18       S   752       37  ...        100.903   \n",
       "2         9       18       S   752       37  ...         14.147   \n",
       "3         9       18       S   752       37  ...         71.268   \n",
       "4         9       18       S   752       37  ...         89.278   \n",
       "\n",
       "   PERCENTGROWTHNOTZ TESTVALUE CONTROLVALUE    TZVALUE  EXPECTEDGROWTH  SCORE  \\\n",
       "0             88.159  332168.0   376781.548  58592.854          90.342    4.0   \n",
       "1            100.763  379656.0   376781.548  58592.854          87.130  -14.0   \n",
       "2             27.498  103608.0   376781.548  58592.854          12.739   -1.0   \n",
       "3             75.736  285360.0   376781.548  58592.854          76.397    5.0   \n",
       "4             90.945  342664.0   376781.548  58592.854          73.681  -16.0   \n",
       "\n",
       "   VALID         PANEL  CELLNAME  \n",
       "0      Y  Renal Cancer     786-0  \n",
       "1      Y  Renal Cancer     786-0  \n",
       "2      Y  Renal Cancer     786-0  \n",
       "3      Y  Renal Cancer     786-0  \n",
       "4      Y  Renal Cancer     786-0  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dl_input_df = pd.read_csv('../data-nci/raw_data/nci/nci-raw-data.csv')\n",
    "dl_input_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3ff248d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "int64\n",
      "float64\n"
     ]
    }
   ],
   "source": [
    "print(dl_input_df['SAMPLE1'].dtype)\n",
    "print(dl_input_df['SAMPLE2'].dtype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78950cf4",
   "metadata": {},
   "source": [
    "I suspect the datatypes have an effect on the missing values, how many missing values do we have when looking at `SAMPLE2`?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "fade4810",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SAMPLE1          0\n",
      "SAMPLE2     812961\n",
      "CELLNAME         0\n",
      "SCORE       815031\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(dl_input_df[['SAMPLE1', 'SAMPLE2', 'CELLNAME', 'SCORE']].isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a4bf06b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing values in SAMPLE2: 0\n"
     ]
    }
   ],
   "source": [
    "dl_input_df = dl_input_df.dropna()\n",
    "dl_input_df['SAMPLE2'] = dl_input_df['SAMPLE2'].astype('int64')\n",
    "print(\"Missing values in SAMPLE2:\", dl_input_df['SAMPLE2'].isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "96f771c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(59898, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE1</th>\n",
       "      <th>SAMPLE2</th>\n",
       "      <th>CELLNAME</th>\n",
       "      <th>SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59893</th>\n",
       "      <td>290</td>\n",
       "      <td>31</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59894</th>\n",
       "      <td>290</td>\n",
       "      <td>46</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59895</th>\n",
       "      <td>290</td>\n",
       "      <td>52</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-7.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59896</th>\n",
       "      <td>290</td>\n",
       "      <td>91</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59897</th>\n",
       "      <td>290</td>\n",
       "      <td>203</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>3.222222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE1  SAMPLE2 CELLNAME     SCORE\n",
       "59893      290       31   HCT-15 -2.444444\n",
       "59894      290       46   HCT-15 -2.000000\n",
       "59895      290       52   HCT-15 -7.888889\n",
       "59896      290       91   HCT-15 -2.444444\n",
       "59897      290      203   HCT-15  3.222222"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dl_input_df = dl_input_df.groupby(['SAMPLE1', 'SAMPLE2', 'CELLNAME']).agg({'SCORE':'mean'}).reset_index()\n",
    "print(dl_input_df.shape)\n",
    "dl_input_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "fb2df9f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(59898, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE1</th>\n",
       "      <th>SAMPLE2</th>\n",
       "      <th>CELLNAME</th>\n",
       "      <th>SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59893</th>\n",
       "      <td>290</td>\n",
       "      <td>31</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59894</th>\n",
       "      <td>290</td>\n",
       "      <td>46</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59895</th>\n",
       "      <td>290</td>\n",
       "      <td>52</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-7.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59896</th>\n",
       "      <td>290</td>\n",
       "      <td>91</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-2.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59897</th>\n",
       "      <td>290</td>\n",
       "      <td>203</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>3.222222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE1  SAMPLE2 CELLNAME     SCORE\n",
       "59893      290       31   HCT-15 -2.444444\n",
       "59894      290       46   HCT-15 -2.000000\n",
       "59895      290       52   HCT-15 -7.888889\n",
       "59896      290       91   HCT-15 -2.444444\n",
       "59897      290      203   HCT-15  3.222222"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dl_input_deletion_list = []\n",
    "dl_input_df = dl_input_df.fillna('missing')\n",
    "for row in dl_input_df.itertuples():\n",
    "    if row[1] == 'missing' or row[2] == 'missing':\n",
    "        dl_input_deletion_list.append(row[0])\n",
    "dl_input_df = dl_input_df.drop(dl_input_df.index[dl_input_deletion_list]).reset_index(drop=True)\n",
    "print(dl_input_df.shape)\n",
    "dl_input_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d6915e4",
   "metadata": {},
   "source": [
    "## Let's put a stop here\n",
    "\n",
    "This seems a bit futile without having input from the authors.\n",
    "\n",
    "## Let's look at the filtered data instead\n",
    "\n",
    "To do that we copy out the data folders from the DSF repository into our own. <br>\n",
    "It can be found under `data-nci` don't forget to modify the `.gitignore` so we don't push around much data. <br>\n",
    "\n",
    "We can ignore the following folders for now: <br>\n",
    "```\n",
    "filtered_data-original\n",
    "plot\n",
    "result\n",
    "```\n",
    "\n",
    "The readme suggests us running the following function to run the model:\n",
    "```\n",
    "python geo_tmain_webgnn.py \n",
    "```\n",
    "this file accepts custom specifications for our GNN <br>\n",
    "More importantly we are \"building\" our model and using the data here:\n",
    "```python\n",
    "def build_geowebgnn_model(args, device, dataset):\n",
    "    print('--- BUILDING UP WEBGNN MODEL ... ---')\n",
    "    # GET PARAMETERS\n",
    "    # [num_gene, num_drug, (adj)node_num]\n",
    "    final_annotation_gene_df = pd.read_csv('./' + dataset + '/filtered_data/kegg_gene_annotation.csv')\n",
    "    gene_name_list = list(final_annotation_gene_df['kegg_gene'])\n",
    "    num_gene = len(gene_name_list)\n",
    "    drug_num_dict_df = pd.read_csv('./' + dataset + '/filtered_data/drug_num_dict.csv')\n",
    "    drug_dict = dict(zip(drug_num_dict_df.Drug, drug_num_dict_df.drug_num))\n",
    "    num_drug = len(drug_dict)\n",
    "    node_num = num_gene + num_drug\n",
    "    # [num_gene_edge, num_drug_edge]\n",
    "    gene_num_df = pd.read_csv('./' + dataset + '/filtered_data/kegg_gene_num_interaction.csv')\n",
    "    gene_num_df = gene_num_df.drop_duplicates()\n",
    "    drugbank_num_df = pd.read_csv('./' + dataset + '/filtered_data/final_drugbank_num_sym.csv')\n",
    "    num_gene_edge = gene_num_df.shape[0]\n",
    "    num_drug_edge = drugbank_num_df.shape[0]\n",
    "    num_edge = num_gene_edge + num_drug_edge\n",
    "    # import pdb; pdb.set_trace()\n",
    "    # BUILD UP MODEL\n",
    "    model = WeBGNNDecoder(input_dim=args.input_dim, hidden_dim=args.hidden_dim, embedding_dim=args.output_dim, \n",
    "                decoder_dim=args.decoder_dim, node_num=node_num, num_edge=num_edge, num_gene_edge=num_gene_edge, device=device)\n",
    "    model = model.to(device)\n",
    "    return model\n",
    "```\n",
    "Here we use: <br>\n",
    "`/filtered_data/kegg_gene_annotation.csv` which is a `.csv` with only 1 column and *2016* gene names <br>\n",
    "to get the number of the genes <br>\n",
    "`/filtered_data/drug_num_dict.csv` which is a `.csv` with two columns,<br> \n",
    "one the names of the drugs <br>\n",
    "and the other their number, going from 2017 (continuing from the genes?) to 2034 <br>\n",
    "so *18* genes <br>\n",
    "to get the number of the genes.<br>\n",
    "Then these numbers are added and provided to the `WeBGNNDecoder` as `node_num`<br>\n",
    "`/filtered_data/kegg_gene_num_interaction.csv` which is a `.csv` with two columns,<br> \n",
    "one the number of the gene<br>\n",
    "the number of the gene it connects to <br>\n",
    "for 2016 genes these are *18512* connections <br>\n",
    "to get the number of the connections (2000, 2010 and 2010, 2000 count as 2 connections).<br>\n",
    "`/filtered_data/final_drugbank_num_sym.csv` which is a `.csv` with two columns,<br> \n",
    "one the number of the drug<br>\n",
    "the number of the gene it connects to <br>\n",
    "for 2016 genes these are *98* connections <br>\n",
    "to get the number of the connections (2019, 1 and 1, 2019 count as 2 connections).<br>\n",
    "Then these numbers are added and provided to the `WeBGNNDecoder` as `edge_num`<br>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f2be821",
   "metadata": {},
   "source": [
    "#### Quick Check: Are some nodes unconnected?\n",
    "\n",
    "No"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f67c5ba5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing numbers: []\n"
     ]
    }
   ],
   "source": [
    "missing_numbers = [\n",
    "    num for num in range(1, 2017) \n",
    "    if not ((gene_num_df['src'] == num).any() or \n",
    "            (gene_num_df['dest'] == num).any())\n",
    "]\n",
    "print(\"Missing numbers:\", missing_numbers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ae18eec",
   "metadata": {},
   "source": [
    "#### Quick Check: Are the edges bidirectional?\n",
    "\n",
    "No, as evident from the code below the num_edges should count edges in both directions seperately, we will see if it is supposed to be this way later. For now I will remebmer this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3726ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e66dad77",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_num_df = pd.read_csv('../data-nci/filtered_data/kegg_gene_num_interaction.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5c1ba67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Duplicate connections (including mirrors):\n",
      "        src  dest\n",
      "18356  2007  2000\n",
      "18470  2008  2000\n",
      "18357  2008  2000\n",
      "18488  2010  2000\n",
      "18358  2010  2000\n",
      "\n",
      "Total duplicate pairs: 662\n"
     ]
    }
   ],
   "source": [
    "mirrored_df = gene_num_df.rename(columns={'src':'dest', 'dest':'src'})\n",
    "\n",
    "# Combine original and mirrored DataFrames\n",
    "combined = pd.concat([gene_num_df, mirrored_df])\n",
    "\n",
    "# Find duplicates (including mirrored pairs)\n",
    "duplicates = combined[combined.duplicated(keep=False)]  # keep=False marks all duplicates\n",
    "\n",
    "# Show duplicates\n",
    "print(\"Duplicate connections (including mirrors):\")\n",
    "print(duplicates.sort_values(by=['src', 'dest']).tail())\n",
    "\n",
    "# Count duplicates\n",
    "print(f\"\\nTotal duplicate pairs: {len(duplicates)//2}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c9e6fb6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18512, 2)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_num_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fb78f614",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18512, 2)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_num_df_cleaned = gene_num_df.drop_duplicates()\n",
    "gene_num_df_cleaned.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89fb73bc",
   "metadata": {},
   "source": [
    "A quick check in the `.csv` confirms the result and `drop_duplicates()` does nothing in this dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7b036b3",
   "metadata": {},
   "source": [
    "### Now Let's go to the `WeBGNNDecoder` to see how it is used\n",
    "\n",
    "It's not used in the `WeBGNNDecoder` at all!!! <br>\n",
    "Which is what is expected from DNN but I am wondering where the multi-omic data comes in? <br>\n",
    "\n",
    "#### Moving on we want to find the next place where we use the data\n",
    "\n",
    "```python\n",
    "def train_geowebgnn(args, fold_n, load_path, iteration_num, device, dataset):\n",
    "    # TRAINING DATASET BASIC PARAMETERS\n",
    "    # [num_feature, num_gene, num_drug]\n",
    "    num_feature = 4\n",
    "    dict_drug_num = pd.read_csv('./' + dataset + '/filtered_data/drug_num_dict.csv')\n",
    "    num_drug = dict_drug_num.shape[0]\n",
    "    final_annotation_gene_df = pd.read_csv('./' + dataset + '/filtered_data/kegg_gene_annotation.csv')\n",
    "    num_gene = final_annotation_gene_df.shape[0]\n",
    "    form_data_path = './' + dataset + '/form_data'\n",
    "```\n",
    "\n",
    "It appears to be later in the same file, in our training loop.\n",
    "Here we use `/form_data` which is not included in the DSF GitHub.\n",
    "\n",
    "### How is `/form_data` made and what is it?\n",
    "\n",
    "```python\n",
    "    xTr = np.load('./' + dataset + '/form_data/xTr' + str(fold_n) + '.npy')\n",
    "    yTr = np.load('./' + dataset + '/form_data/yTr' + str(fold_n) + '.npy')\n",
    "    drugTr =  np.load('./' + dataset + '/form_data/drugTr' + str(fold_n) + '.npy')\n",
    "    edge_index = torch.from_numpy(np.load(form_data_path + '/edge_index.npy')).long() \n",
    "```\n",
    "\n",
    "##### It seems to be our training data!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e2a9ef7",
   "metadata": {},
   "source": [
    "#### First we Randomize our input\n",
    "\n",
    "```python\n",
    "def input_random(dataset):\n",
    "    final_input_df = pd.read_csv('./' + dataset + '/filtered_data/final_dl_input.csv')\n",
    "    random_final_input_df = final_input_df.sample(frac = 1)\n",
    "    random_final_input_df.to_csv('./' + dataset + '/filtered_data/random_final_dl_input.csv', index = False, header = True)\n",
    "```\n",
    "This just shuffles our data\n",
    "\n",
    "#### Then we create our folds\n",
    "\n",
    "```python\n",
    "def split_k_fold(k, dataset):\n",
    "    random_final_dl_input_df = pd.read_csv('./' + dataset + '/filtered_data/random_final_dl_input.csv')\n",
    "    num_points = random_final_dl_input_df.shape[0]\n",
    "    num_div = int(num_points / k)\n",
    "    num_div_list = [i * num_div for i in range(0, k)]\n",
    "    num_div_list.append(num_points)\n",
    "    # SPLIT [RandomFinal_NCI60_DeepLearningInput] INTO [k] FOLDS\n",
    "    for place_num in range(k):\n",
    "        low_idx = num_div_list[place_num]\n",
    "        high_idx = num_div_list[place_num + 1]\n",
    "        print('\\n--------TRAIN-TEST SPLIT WITH TEST FROM ' + str(low_idx) + ' TO ' + str(high_idx) + '--------')\n",
    "        split_input_df = random_final_dl_input_df[low_idx : high_idx]\n",
    "        split_input_df.to_csv('./' + dataset + '/filtered_data/split_input_' + str(place_num + 1) + '.csv', index = False, header = True)\n",
    "```\n",
    "This creates equally big splits as `.csv` files. <br>\n",
    "#### BUT the `.csv` files from their Repo are faulty\n",
    "split_1 contains most of the samples and split_2 almost none.<br>\n",
    "they have it correct in another folder (`filtered_data-original`) but I am not sure which one they used<br>\n",
    "\n",
    "#### There seems to be no other cross validation except leave drug combination out!\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5fd2ee63",
   "metadata": {},
   "outputs": [],
   "source": [
    "whole_set = pd.read_csv('../data-nci/filtered_data/random_final_dl_input.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "55f3e4ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug A</th>\n",
       "      <th>Drug B</th>\n",
       "      <th>Cell Line Name</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Dasatinib</td>\n",
       "      <td>Docetaxel</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>-4.777778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Streptozocin</td>\n",
       "      <td>Tretinoin</td>\n",
       "      <td>HOP-62</td>\n",
       "      <td>-12.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Thalidomide</td>\n",
       "      <td>Paclitaxel</td>\n",
       "      <td>NCI-H322M</td>\n",
       "      <td>1.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mitotane</td>\n",
       "      <td>Sirolimus</td>\n",
       "      <td>SN12C</td>\n",
       "      <td>-7.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sirolimus</td>\n",
       "      <td>Cladribine</td>\n",
       "      <td>NCI-H322M</td>\n",
       "      <td>-18.555556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3937</th>\n",
       "      <td>Vorinostat</td>\n",
       "      <td>Nilotinib</td>\n",
       "      <td>DU-145</td>\n",
       "      <td>-4.555556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3938</th>\n",
       "      <td>Paclitaxel</td>\n",
       "      <td>Nilotinib</td>\n",
       "      <td>SF-268</td>\n",
       "      <td>-20.111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3939</th>\n",
       "      <td>Mitotane</td>\n",
       "      <td>Nilotinib</td>\n",
       "      <td>CCRF-CEM</td>\n",
       "      <td>-8.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3940</th>\n",
       "      <td>Everolimus</td>\n",
       "      <td>Romidepsin</td>\n",
       "      <td>SW-620</td>\n",
       "      <td>0.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3941</th>\n",
       "      <td>Fulvestrant</td>\n",
       "      <td>Dasatinib</td>\n",
       "      <td>SNB-75</td>\n",
       "      <td>6.222222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3942 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Drug A      Drug B Cell Line Name      Score\n",
       "0        Dasatinib   Docetaxel         HCT-15  -4.777778\n",
       "1     Streptozocin   Tretinoin         HOP-62 -12.000000\n",
       "2      Thalidomide  Paclitaxel      NCI-H322M   1.888889\n",
       "3         Mitotane   Sirolimus          SN12C  -7.666667\n",
       "4        Sirolimus  Cladribine      NCI-H322M -18.555556\n",
       "...            ...         ...            ...        ...\n",
       "3937    Vorinostat   Nilotinib         DU-145  -4.555556\n",
       "3938    Paclitaxel   Nilotinib         SF-268 -20.111111\n",
       "3939      Mitotane   Nilotinib       CCRF-CEM  -8.444444\n",
       "3940    Everolimus  Romidepsin         SW-620   0.222222\n",
       "3941   Fulvestrant   Dasatinib         SNB-75   6.222222\n",
       "\n",
       "[3942 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "whole_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9399c4d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Duplicate connections (including mirrors):\n",
      "Empty DataFrame\n",
      "Columns: [Drug A, Drug B, Cell Line Name, Score]\n",
      "Index: []\n",
      "\n",
      "Total duplicate pairs: 0\n"
     ]
    }
   ],
   "source": [
    "mirrored_whole_set = whole_set.rename(columns={'Drug A':'Drug B', 'Drug B':'Drug A'})\n",
    "\n",
    "# Combine original and mirrored DataFrames\n",
    "combined_wh = pd.concat([whole_set, mirrored_whole_set])\n",
    "\n",
    "# Find duplicates (including mirrored pairs)\n",
    "duplicates_wh = combined_wh[combined_wh.duplicated(keep=False)]  # keep=False marks all duplicates\n",
    "\n",
    "# Show duplicates\n",
    "print(\"Duplicate connections (including mirrors):\")\n",
    "print(duplicates_wh.sort_values(by=['Drug A', 'Drug B']).tail())\n",
    "\n",
    "# Count duplicates\n",
    "print(f\"\\nTotal duplicate pairs: {len(duplicates_wh)//2}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "792e3887",
   "metadata": {},
   "source": [
    "No duplicates or Drug A Drug B leakage here!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d95409c7",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0171d392",
   "metadata": {},
   "source": [
    "# CONTINUATION IN `create-data.ipynb`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
